### PR TITLE
Refactor helidon config 4.0

### PR DESCRIPTION
**Pull request by trentjeff**
Fix for #4764

### CODE REVIEW

**Comment by romain-grecourt** [comment]
Can we separate pico into its own dedicated PR ?

File: pico/LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Comment (romain-grecourt) [inline pico/LICENSE.txt:1]: There is already a LICENSE file at the root of the repo. Is this intended ?

File: config/config/pom.xml [hunk 2]
```diff
@@ -42,6 +42,10 @@
             <groupId>jakarta.annotation</groupId>
             <artifactId>jakarta.annotation-api</artifactId>
         </dependency>
+        <dependency>
```
Comment (romain-grecourt) [inline config/config/pom.xml:45]: This feels backwards.

The declarative API should be on layered on top of the programmatic API, right ?
As it stands `helidon-config` pulls `pico-config` even if pico isn't used.

File: pico/config/pom.xml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    <groupId>io.helidon.pico.config</groupId>
+    <artifactId>helidon-pico-config-project</artifactId>
+    <name>Helidon Pico Config</name>
+    <packaging>pom</packaging>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
+
+        <!--        <maven.deploy.skip>true</maven.deploy.skip>-->
```
Comment (romain-grecourt) [inline pico/config/pom.xml:40]: Commented-out.

File: pico/config/pom.xml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        <!--        <maven.sources.skip>true</maven.sources.skip>-->
+        <!--        <maven.javadoc.skip>true</maven.javadoc.skip>-->
+        <!--        <spotbugs.skip>true</spotbugs.skip>-->
+        <!--        <dependency-check.skip>true</dependency-check.skip>-->
+        <!--        <helidon.services.skip>true</helidon.services.skip>-->
+    </properties>
+
+    <modules>
+<!--        <module>api</module>-->
```
Comment (romain-grecourt) [inline pico/config/pom.xml:49]: Commented-out.

File: pico/pom.xml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    <groupId>io.helidon.pico</groupId>
+    <artifactId>helidon-pico-project</artifactId>
+    <name>Helidon Pico Project</name>
+
+    <packaging>pom</packaging>
+
+    <properties>
+        <version.java>11</version.java>
+
+        <!--        &lt;!&ndash; javadoc fails on Java 19 &ndash;&gt;-->
```
Comment (romain-grecourt) [inline pico/pom.xml:41]: Commented-out.

File: pico/pom.xml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        <!--        <maven.javadoc.skip>true</maven.javadoc.skip>-->
+
+<!--        <javax.injection.version>1</javax.injection.version>-->
```
Comment (romain-grecourt) [inline pico/pom.xml:44]: Commented-out.

File: integrations/cdi/datasource/pom.xml [hunk 7]
```diff
@@ -47,6 +47,11 @@
             <artifactId>jakarta.annotation-api</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (romain-grecourt) [inline integrations/cdi/datasource/pom.xml:50]: Not sure that belongs in this PR.

Comment (trentjeff) [reply integrations/cdi/datasource/pom.xml:50]: it was required to build since some of the scopes changed from provided to runtime as part of this change.

File: pico/LICENSE.txt (hunk 1 above)
Comment (trentjeff) [reply pico/LICENSE.txt:1]: There was a point in time where pico was going to be a top-level module. Will discuss it with @tomas-langer - currently it is a child of helidon (obviously), but now that we are close to having our our org on gitHub this might move.

Comment (romain-grecourt) [reply pico/LICENSE.txt:1]: Since you are adding pico as a module as part of this PR, the license file should be removed.
You can add the license file back If/when pico moves into its own repository.

File: config/config/pom.xml (hunk 2 above)
Comment (tomas-langer) [reply config/config/pom.xml:45]: There is a chicken-egg problem with config. It can be solved two ways:
1. create a config API that is implemented by `helidon-config` and backward compatible
2. make `helidon-config` the config API and have a separate implementation module (not backward compatible)

The first option seems to have the least impact on Helidon

File: pico/api/pom.xml [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>io.helidon.common</groupId>
+            <artifactId>helidon-common</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>jakarta.inject</groupId>
+            <artifactId>jakarta.inject-api</artifactId>
+            <scope>compile</scope>
```
Comment (wojtask9) [inline pico/api/pom.xml:41]: I don't understand idea of this.
Starting with 4.x jakarta.* annotations will be required for SE ?

Comment (trentjeff) [reply pico/api/pom.xml:41]: The jakarta.inject module is needed (as compile scope) for the definition of the Provider class, sadly in the same package with other annotations that would otherwise only be needed as provided scope.

File: common/config/pom.xml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
```
[the remark this answers is not in the document]
Comment (trentjeff) [reply common/config/pom.xml:38]: resolved & removed/pushed

File: common/common/pom.xml [hunk 10]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
 
-    Copyright (c) 2017, 2022 Oracle and/or its affiliates.
+    Copyright (c) 2022 Oracle and/or its affiliates.
```
Comment (tomas-langer) [inline common/common/pom.xml:4]: This is wrong copyright change. It must be `2017, 2022`. Please revert this change

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+ * limitations under the License.
+ */
+
+package io.helidon.common.config;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * <h2>Configuration</h2>
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:23]: Adding headers to javadocs is tricky and not really needed here, when we have a single sentence description

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+     * <p>
+     * The fully-qualified key is a sequence of tokens derived from the name of
+     * each node along the path from the config root to the current node. Tokens
+     * are separated by {@code .} (the dot character). See {@link #name()} for
+     * more information on the format of each token.
+     *
+     * @return current config node key
+     * @see #name()
+     */
+    BaseKey key();
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:41]: BaseKey? This should return `Key`. Our API should be `Key`, `BaseKey` I would expect to be an abstract class I can extend, not API

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+    /**
+     * Returns {@code true} if this node exists and is a leaf node (has no
+     * children).
+     * <p>
+     * A leaf node has no nested configuration subtree and has a single value.
+     *
+     * @return {@code true} if the node is existing leaf node, {@code false}
+     *         otherwise.
+     */
+    boolean isLeaf();
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:164]: Not really needed. This is basically `!isObject() && !isKey() && hasValue()`.
Also method `hasValue()` is missing

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+
+    /**
+     * Returns list of specified type.
+     *
+     * @param type type class
+     * @param <T>  type of list elements
+     * @return a typed list with values
+     * @throws io.helidon.common.config.ConfigException in case of problem to map property value.
+     */
+    default <T> ConfigValue<List<T>> asList(Class<T> type) throws ConfigException {
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:208]: This should not have a default implementation. Esp. not an implementation like this

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+        throw new ConfigException("unsupported; do you have a valid config module in the classpath?");
+    }
+
+    /**
+     * Transform all leaf nodes (values) into Map instance.
+     *
+     * @return new Map instance that contains all config leaf node values
+     * @throws io.helidon.common.config.ConfigException in case the node is Type#MISSING.
+     */
+    default ConfigValue<Map<String, String>> asMap() throws ConfigException {
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:218]: dtto

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+     *  key-token = *( unescaped / escaped )
+     *  unescaped = %x00-2D / %x2F-7D / %x7F-10FFFF
+     *            ; %x2E ('.') and %x7E ('~') are excluded from 'unescaped'
+     *    escaped = "~" ( "0" / "1" )
+     *            ; representing '~' and '.', respectively
+     * }</pre>
+     *
+     * @see Config#key()
+     */
+    interface BaseKey extends Comparable<BaseKey> {
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:246]: This should be named `Key`. BaseKey would mean an abstract class. Also this is one of the main APIs of Config, so not prefixes should be used.

File: common/config/src/main/java/io/helidon/common/config/Config.java [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+         * Returns instance of Key that represents key of parent config node.
+         * <p>
+         * If the key represents root config node it throws an exception.
+         *
+         * @return key that represents key of parent config node.
+         * @see #isRoot()
+         * @throws IllegalStateException in case you attempt to call this method on a root node
+         * @throws io.helidon.common.config.ConfigException if not defined
+         */
+        default BaseKey parent() throws ConfigException {
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/Config.java:292]: Do not add default implementations like this

File: common/config/src/main/java/io/helidon/common/config/package-info.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2017, 2022 Oracle and/or its affiliates.
```
Comment (tomas-langer) [inline common/config/src/main/java/io/helidon/common/config/package-info.java:2]: Copyright should be just `2022`, this file did not exist in 2017

File: common/pom.xml [hunk 19]
```diff
@@ -32,8 +32,9 @@
 
     <modules>
         <module>common</module>
-        <module>reactive</module>
+        <module>config</module>
         <module>configurable</module>
+        <module>reactive</module>
```
Comment (tomas-langer) [inline common/pom.xml:37]: why is this reordered? Please simplify the change

File: config/config/etc/spotbugs/exclude.xml [hunk 20]
```diff
@@ -83,5 +83,20 @@
         <Method name="parse"/>
         <Bug pattern="RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE"/>
     </Match>
+    <Match>
+        <!-- Path shadows common/config -->
```
Comment (tomas-langer) [inline config/config/etc/spotbugs/exclude.xml:87]: what does `Path` mean in these descriptions? I do not understand this explanation

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: common/config/src/main/java/io/helidon/common/config/Config.java (hunk 13 above)
Comment (trentjeff) [reply common/config/src/main/java/io/helidon/common/config/Config.java:164]: There is no isKey() on either common/config/Config or config/config.  I will add hasValue() decl on common/config.

Comment (tomas-langer) [reply common/config/src/main/java/io/helidon/common/config/Config.java:159]: sorry, meant `isList()` not `isKey()`

**Review by tomas-langer — APPROVED** [review]
LGTM
